x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java

```diff
@@ -173,23 +182,79 @@ private void checkForMissingDataIfNecessary() {
+                    totalRecordsMissing);
+
+                // Have we an annotation that covers the same area with the same message?
+                // Cannot use annotation.equals(other) as that checks createTime
+                if (lastDataCheckAnnotation != null
+                    && annotation.getAnnotation().equals(lastDataCheckAnnotation.getAnnotation())
+                    && annotation.getTimestamp().equals(lastDataCheckAnnotation.getTimestamp())
+                    && annotation.getEndTimestamp().equals(lastDataCheckAnnotation.getTimestamp())) {
+                    return;
+                }
+
+                if (lastDataCheckAnnotationId != null) {
+                    updateAnnotation(annotation);
+                } else {
+                    lastDataCheckAnnotationId = addAndSetDelayedDataAnnotation(annotation);
                 }
             }
         }
     }
 
+    private Annotation createAnnotation(Date startTime, Date endTime, long recordsMissing) {
+       String msg = Messages.getMessage(Messages.JOB_AUDIT_DATAFEED_MISSING_DATA, recordsMissing,
+            XContentElasticsearchExtension.DEFAULT_DATE_PRINTER.print(endTime.getTime()));
+       return new Annotation(msg,
+           new Date(currentTimeSupplier.get()),
+           "ml-delayed-data-checker",
+           startTime,
+           endTime,
+           jobId,
+           null,
+           null,
+           "annotation");
+    }
+
+    private String addAndSetDelayedDataAnnotation(Annotation annotation) {
+        try (XContentBuilder xContentBuilder = annotation.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS)) {
+            IndexRequest request = new IndexRequest(AnnotationIndex.WRITE_ALIAS_NAME);
+            request.source(xContentBuilder);
+            IndexResponse response = client.index(request).actionGet();
+            lastDataCheckAnnotation = annotation;
+            return response.getId();
+        } catch (IOException ex) {
+            String errorMessage = "[" + jobId + "] failed to create annotation for delayed data checker.";
+            LOGGER.error(errorMessage, ex);
+            auditor.error(jobId, errorMessage);
+            return null;
+        }
+    }
+
+    private void updateAnnotation(Annotation annotation) {
+        Annotation updatedAnnotation = new Annotation(lastDataCheckAnnotation);
+        updatedAnnotation.setModifiedUsername("ml-delayed-data-checker");
+        updatedAnnotation.setModifiedTime(new Date(currentTimeSupplier.get()));
+        updatedAnnotation.setAnnotation(annotation.getAnnotation());
+        updatedAnnotation.setTimestamp(annotation.getTimestamp());
+        updatedAnnotation.setEndTimestamp(annotation.getEndTimestamp());
+        try (XContentBuilder xContentBuilder = updatedAnnotation.toXContent(XContentFactory.jsonBuilder(), ToXContent.EMPTY_PARAMS)) {
+            UpdateRequest updateRequest = new UpdateRequest(AnnotationIndex.WRITE_ALIAS_NAME, lastDataCheckAnnotationId);
```

Comment:
It might be best to just use an index request here as we have the full document. By using the same ID it will overwrite assuming it’s going to the same index. In future the write alias could have rolled so this will be writing to a different index and hence not updating. (At that time clients will have to deduplicate after searching for annotations.)

Reply:
@droberts195 what about upsert?

Reply:
@droberts195 I added `updateRequest.docAsUpsert(true);` since I am copying over the entire doc and setting it appropriately, we can just use its contents as an upsert if the document with the stored ID does not exist.

Reply:
From https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update.html:

> To fully replace the existing document, the index API should be used instead.

So I think a simple index would be more efficient than an upsert.

Reply:
Seems logical to me :D.